File: src/main/java/it/alessandroiezzi/genwebsite/Page.java
/*
 * Copyright (C) 2021  Alessandro Iezzi dev@alessandroiezzi.it
 *
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package it.alessandroiezzi.genwebsite;

import com.vladsch.flexmark.html.HtmlRenderer;
import com.vladsch.flexmark.parser.Parser;
import com.vladsch.flexmark.util.ast.Node;
import com.vladsch.flexmark.util.data.MutableDataSet;
import lombok.Getter;
import lombok.Setter;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Page {
    @Setter private String rootDir = "";
    @Getter @Setter private String out = "";
    @Getter @Setter private String content = "";
    @Getter private String in = "";
    @Getter @Setter private String template = "";
    @Setter private String title = "";
    @Getter @Setter private String id = "";
    @Getter @Setter private String dir = "";
    @Getter private boolean parsed = false;
    @Getter private String date;

    public Page() {
        date = new String("");
    }

    public String getTitle() {
        if (!parsed) {
            throw new RuntimeException("You must call parse before getTitle");
        }

        System.out.println("Page title: " + title);

        return title;
    }

    public void setIn(String in) {
        parsed = false;
        this.in = in;
    }

    public void setDate(String date) {
        this.date = date;
        System.out.println("-------------------> " + date);
    }

    public String parse() throws IOException {
        if (parsed)
            return content;

        if (in != null && !in.isEmpty() && !in.trim().isEmpty()) {
            String[] splited = in.split("\\.");
            String extension = splited[splited.length - 1];

            StringBuilder sb = new StringBuilder();
            BufferedReader br = Files.newBufferedReader(Paths.get(rootDir).resolve(in));
            String line;
            while ((line = br.readLine()) != null) {
                if (line.startsWith("# property:")) {
                    String property = line.replace("# property:", "");
                    String[] splitted = property.split("=");
                    if (splitted.length > 1) {
                        switch(splitted[0].trim()) {
                            case "title":
                                title = splitted[1].trim();
                                break;
                            case "date":
                                date = splitted[1].trim();
                                break;
                        }
                    }
                } else {
                    sb.append(line).append("\n");
                }
            }

            switch (extension) {
                case "md":
                    System.out.println("Parsing markdown");
                    MutableDataSet options = new MutableDataSet();

                    // uncomment to set optional extensions
                    //options.set(Parser.EXTENSIONS, Arrays.asList(TablesExtension.create(), StrikethroughExtension.create()));

                    // uncomment to convert soft-breaks to hard breaks
                    //options.set(HtmlRenderer.SOFT_BREAK, "<br />\n");

                    Parser parser = Parser.builder(options).build();
                    HtmlRenderer renderer = HtmlRenderer.builder(options).build();

                    // You can re-use parser and renderer instances
                    Node document = parser.parse(sb.toString());

                    content = renderer.render(document);
                    break;
                case "html":
                case "htm":
                case "xhtm":
                    content = sb.toString();
                    break;
            }
        }

        parsed = true;

        return content;
    }
}

File: page.rb
# A static website generator
# Copyright (C) 2022  Alessandro Iezzi <aiezzi AT alessandroiezzi DOT it>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

require "erb"
require "optparse"
require "yaml"
require 'date'

class Page
  attr_reader :pageFileName

	attr :fileName
	attr :title
	attr :pageNames
	attr :description
	attr :date
	attr :classes
	attr :category
	attr :master
	attr :baseHref


	def parseConfig(configFile)
		if (File.exist?(configFile))
			config = YAML.load_file(configFile)
			if (config != nil)
				@title       = config['title']
				@pageNames   = config['pageNames']
				@description = config['description']
				@classes     = config['classes']
				@master      = config['master']
				@category    = config['category']
				if (@date == nil)
					@date = config['date']
				end
				if (@baseHref == nil)
					@baseHref= config['baseHref']
				end
			end
		end
	end

	attr_writer :fileName
	attr_writer :title
	attr_writer :pageNames
	attr_writer :description
	attr_writer :date
	attr_writer :classes
	attr_writer :category
	attr_writer :master
	attr_writer :content
	attr_writer :baseHref

  def initialize1 title, pageFileName, pageNames, description, date, classes, category
    @title = title
    @pageFileName = pageFileName
    @pageNames = pageNames
    @description = description
    @date = date
    @classes = classes
    @category = category
  end

  def render1 path
    content = File.read(File.expand_path(path))
    t = ERB.new(content)
    t.result(binding)
  end
end
